### PR TITLE
[GISEL] G_SPLAT_VECTOR can take a splat that is larger than the vector element

### PR DESCRIPTION
**Pull request by michaelmaitland**
This is what SelectionDAG does. We'd like to reuse SelectionDAG patterns.

### CODE REVIEW

File: llvm/docs/GlobalISel/GenericOpcode.rst [hunk 1]
```diff
@@ -690,6 +690,10 @@ G_SPLAT_VECTOR
 
 Create a vector where all elements are the scalar from the source operand.
 
+The type of the operand must be equal to or smaller than the vector element
+type. If the operand is smaller than the vector element type, the scalar is
+implicitly truncated to the vector element type.
```
Comment (tschuett) [inline llvm/docs/GlobalISel/GenericOpcode.rst:695]: extended?

Comment (michaelmaitland) [reply llvm/docs/GlobalISel/GenericOpcode.rst:695]: I got it all backwards 😆 . The operand should be allowed to be greater than the size of the vector element.

**Comment by tschuett** [comment]
The title still says smaller. Assuming that this is again for patterns, but `G_SPLAT_VECTOR(G_TRUNC(s128))` should suffice during legalisation?

**Comment by michaelmaitland** [comment]
> Assuming that this is again for patterns, but G_SPLAT_VECTOR(G_TRUNC(s128)) should suffice during legalisation?

This is for patterns and for equivalence to SelectionDAG ISD::SPLAT_VECTOR.

**Comment by tschuett** [comment]
I am again confused. You want to take i64 splats on 32-bit machines and truncate the i64 to i32? You can just state that i64 are not legal?!?

**Comment by michaelmaitland** [comment]
> I am again confused. You want to take i64 splats on 32-bit machines and truncate the i64 to i32? You can just state that i64 are not legal?!?

From selection dag:
> The type of the operand must match the vector element type, except when they are integer types. In this case the operand is allowed to be wider than the vector element type, and is implicitly truncated to it.

I want to do this since the patterns have an sXLen operand which can be s32 on RV32 and I should be able to create a 8b, 16b splat vector.

File: llvm/lib/CodeGen/MachineVerifier.cpp [hunk 2]
```diff
@@ -1768,16 +1768,23 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
...
-    if (!SrcTy.isScalar())
+    if (!SrcTy.isScalar()) {
       report("Source type must be a scalar", MI);
+      break;
+    }
 
-    if (DstTy.getScalarType() != SrcTy)
-      report("Element type of the destination must be the same type as the "
-             "source type",
+    if (TypeSize::isKnownGT(DstTy.getScalarType().getSizeInBits(),
```
Comment (arsenm) [inline llvm/lib/CodeGen/MachineVerifier.cpp:1781]:
```suggestion
    if (TypeSize::isKnownGT(DstTy.getElementType().getSizeInBits(),
```

**Review by aemerson — APPROVED** [review]
This seems reasonable to me if it allows us to import more patterns.

File: llvm/lib/CodeGen/MachineVerifier.cpp (hunk 2 above)
Comment (michaelmaitland) [reply llvm/lib/CodeGen/MachineVerifier.cpp]: updated.

**Pull request merged**